components/ContainedInputViews/src/MDCContainedInputView.h

```diff
@@ -0,0 +1,319 @@
+   */
+  MDCContainedInputViewStateFocused = 1 << 1,
+  /**
+   This state most closely resembles the @c selected UIControlState.
+   */
+  MDCContainedInputViewStateActivated = 1 << 2,
+  /**
+   The error state.
+   */
+  MDCContainedInputViewStateErrored = 1 << 3,
+  /**
+   The disabled state.
+   */
+  MDCContainedInputViewStateDisabled = 1 << 4,
+};
+
+/**
+ Dictates the relative importance of the underline labels, and the order in which they are laid out.
+ */
+typedef NS_ENUM(NSUInteger, MDCContainedInputViewUnderlineLabelDrawPriority) {
+  /**
+   When the priority is @c .leading, the @c leadingUnderlineLabel will be laid out first within the
+   horizontal space available for @b both underline labels. Any remaining space will then be given
+   for the @c trailingUnderlineLabel.
+   */
+  MDCContainedInputViewUnderlineLabelDrawPriorityLeading,
+  /**
+   When the priority is @c .trailing, the @c trailingUnderlineLabel will be laid out first within
+   the horizontal space available for @b both underline labels. Any remaining space will then be
+   given for the @c leadingUnderlineLabel.
+   */
+  MDCContainedInputViewUnderlineLabelDrawPriorityTrailing,
+  /**
+   When the priority is @c .custom, the @c customUnderlineLabelDrawPriority property will be used to
+   divide the space available for the two underline labels.
+   */
+  MDCContainedInputViewUnderlineLabelDrawPriorityCustom,
+};
+
+/**
+ This enum represents different states the floating label can be in.
+ */
+typedef NS_ENUM(NSUInteger, MDCContainedInputViewFloatingLabelState) {
+  /**
+   The state where the floating label is not visible.
+   */
+  MDCContainedInputViewFloatingLabelStateNone,
+  /**
+   The state where the floating label is floating.
+   */
+  MDCContainedInputViewFloatingLabelStateFloating,
+  /**
+   The state where the floating label is occupying the normal text area.
+   */
+  MDCContainedInputViewFloatingLabelStateNormal,
+};
+
+@protocol MDCContainedInputViewStyle;
+@protocol MDCContainedInputViewColorScheming;
+
+@protocol MDCContainedInputView <NSObject>
+/**
+ Dictates the @c MDCContainerStyle of the text field. Defaults to an instance of
+ MDCContainerStyleBase.
+ */
+@property(nonatomic, strong, nonnull) id<MDCContainedInputViewStyle> containerStyle;
+
+/**
+ Describes the current @c MDCContainedInputViewState of the view.
+ */
+@property(nonatomic, assign, readonly) MDCContainedInputViewState containedInputViewState;
+
+/**
+ Describes the current @c MDCContainedInputViewFloatingLabelState of the contained input view. This
+ value is affected by things like the view's state, the value for @c canFloatingLabelFloat, and the
+ text of the floating label.
+ */
+@property(nonatomic, assign, readonly) MDCContainedInputViewFloatingLabelState floatingLabelState;
+
+/**
+ The @c floatingLabel is a label that occupies the text area when there is no text and that floats
+ above the text once there is some. It is distinct from a placeholder.
```

Comment:
Tip: Can you elaborate on how this is different from a placeholder?

Reply:
I actually only recently learned that these two were distinct... I had to do some refactoring on the text field to reflect this reality haha. See http://b/127976833 for more context on this particular thing